plugins/wacom-usb/fu-wac-firmware.c

```diff
@@ -18,244 +18,264 @@ struct _FuWacFirmware {
 
+#define FU_WAC_FIRMWARE_TOKENS_MAX 100000 /* lines */
+
 typedef struct {
 	guint32 addr;
 	guint32 sz;
 	guint32 prog_start_addr;
 } FuFirmwareWacHeaderRecord;
 
+typedef struct {
+	FuFirmware *firmware;
+	FwupdInstallFlags flags;
+	GPtrArray *header_infos;
+	GString *image_buffer;
+	guint8 images_cnt;
+} FuWacFirmwareTokenHelper;
+
 static gboolean
-fu_wac_firmware_parse(FuFirmware *firmware,
-		      GBytes *fw,
-		      guint64 addr_start,
-		      guint64 addr_end,
-		      FwupdInstallFlags flags,
-		      GError **error)
+fu_wac_firmware_tokenize_cb(GString *token, guint token_idx, gpointer user_data, GError **error)
 {
-	gsize len;
-	guint8 *data;
-	guint8 images_cnt = 0;
-	g_auto(GStrv) lines = NULL;
-	g_autoptr(GPtrArray) header_infos = g_ptr_array_new_with_free_func(g_free);
-	g_autoptr(GString) image_buffer = NULL;
-
-	/* check the prefix (BE) */
-	data = (guint8 *)g_bytes_get_data(fw, &len);
-	if (len < 5 || memcmp(data, "WACOM", 5) != 0) {
-		g_set_error_literal(error,
-				    FWUPD_ERROR,
-				    FWUPD_ERROR_INTERNAL,
-				    "invalid .wac prefix");
-		return FALSE;
-	}
-
-	/* parse each line */
-	lines = fu_common_strnsplit((const gchar *)data, len, "\n", -1);
-	for (guint i = 0; lines[i] != NULL; i++) {
-		g_autofree gchar *cmd = NULL;
-
-		/* remove windows line endings */
-		g_strdelimit(lines[i], "\r", '\0');
-		cmd = g_strndup(lines[i], 2);
-
-		/* ignore blank lines */
-		if (g_strcmp0(cmd, "") == 0)
-			continue;
-
-		/* Wacom-specific metadata */
-		if (g_strcmp0(cmd, "WA") == 0) {
-			gsize cmdlen = strlen(lines[i]);
-
-			/* header info record */
-			if (cmdlen > 3 && memcmp(lines[i] + 2, "COM", 3) == 0) {
-				guint8 header_image_cnt = 0;
-				if (cmdlen != 40) {
-					g_set_error(error,
-						    FWUPD_ERROR,
-						    FWUPD_ERROR_INTERNAL,
-						    "invalid header, got %" G_GSIZE_FORMAT " bytes",
-						    cmdlen);
-					return FALSE;
-				}
-				if (!fu_firmware_strparse_uint4_safe(lines[i],
-								     cmdlen,
-								     5,
-								     &header_image_cnt,
-								     error))
-					return FALSE;
-				for (guint j = 0; j < header_image_cnt; j++) {
-					g_autofree FuFirmwareWacHeaderRecord *hdr = NULL;
-					hdr = g_new0(FuFirmwareWacHeaderRecord, 1);
-					if (!fu_firmware_strparse_uint32_safe(lines[i],
-									      cmdlen,
-									      (j * 16) + 6,
-									      &hdr->addr,
-									      error))
-						return FALSE;
-					if (!fu_firmware_strparse_uint32_safe(lines[i],
-									      cmdlen,
-									      (j * 16) + 14,
-									      &hdr->sz,
-									      error))
-						return FALSE;
-					g_debug("header_fw%u_addr: 0x%x", j, hdr->addr);
-					g_debug("header_fw%u_sz:   0x%x", j, hdr->sz);
-					g_ptr_array_add(header_infos, g_steal_pointer(&hdr));
-				}
-				continue;
+	FuWacFirmwareTokenHelper *helper = (FuWacFirmwareTokenHelper *)user_data;
+	g_autofree gchar *cmd = NULL;
+
+	/* remove WIN32 line endings */
+	g_strdelimit(token->str, "\r\x1a", '\0');
```

Comment:
same comment as other plugin, this is removing the EOF, I'd think you want to remove all `\r`?